store/etcd/types_transformation.go

```diff
@@ -0,0 +1,182 @@
+package etcd
+
+import (
+	"fmt"
+
+	"github.com/heetch/regula"
+	"github.com/heetch/regula/rule"
+	pb "github.com/heetch/regula/store/etcd/proto"
+)
+
+// toProtobufRuleset creates a protobuf Ruleset from a regula.Ruleset.
+func toProtobufRuleset(rs *regula.Ruleset) *pb.Ruleset {
+	pbrs := &pb.Ruleset{
+		Type: rs.Type,
+	}
+
+	for _, r := range rs.Rules {
```

Comment:
:tulip: Since we are talking about performance here, you can preallocate the `pbrs.Rules` slice because you already know the size of the rules. I suggest running the benchmark before and after the change

Reply:
There is no huge difference 🤔

With append:
```
BenchmarkAppendToProto-4          1000000          1667 ns/op        1368 B/op          33 allocs/op
BenchmarkAppendFromProto-4        1000000          1480 ns/op         872 B/op          22 allocs/op
```

Without append:
```
BenchmarkAppendToProto-4          1000000          1534 ns/op        1344 B/op          30 allocs/op
BenchmarkAppendFromProto-4        1000000          1388 ns/op         864 B/op          21 allocs/op
```

The following are the benchmark functions:

```go
func BenchmarkAppendToProto(b *testing.B) {
	rs, err := regula.NewBoolRuleset(
		rule.New(rule.And(rule.Not(rule.BoolValue(false)), rule.BoolParam("param")), rule.BoolValue(true)),
		rule.New(rule.And(rule.BoolParam("1st-param"), rule.BoolParam("2nd-param")), rule.BoolValue(false)),
	)
	require.NoError(b, err)

	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		_ = rulesetToProtobuf(rs)
	}
}
```

```go
func BenchmarkAppendFromProto(b *testing.B) {
	rs, err := regula.NewBoolRuleset(
		rule.New(rule.And(rule.Not(rule.BoolValue(false)), rule.BoolParam("param")), rule.BoolValue(true)),
		rule.New(rule.And(rule.BoolParam("1st-param"), rule.BoolParam("2nd-param")), rule.BoolValue(false)),
	)
	require.NoError(b, err)

	pbrs := rulesetToProtobuf(rs)

	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		_ = rulesetFromProtobuf(pbrs)
	}
}
```

Reply:
Done in https://github.com/heetch/regula/pull/86/commits/76f07b47bf6d74ba52122fc8bb4236afba361c8c.

Reply:
There won't be any with only two rules, try to generate 10 or more, the difference should be more visible.